### PR TITLE
JSDK 2337: media monitor

### PR DESCRIPTION
**Pull request by makarandp0**
This change adds Ice media monitor - to detect the disconnects, and start reconnect early. This is mostly based on corresponding CPP work in https://github.com/twilio/rtc-cpp/pull/640

Change:
1) added a new class IceMediaMonitor `lib/signaling/v2/icemediamonitor.js` that queries the peerConnection for media stats, and calls back when it detects inactivity. (inactivity is defined as no bytes received in 3 seconds)
2) PeerConnection (lib/signaling/v2/peerconnection.js) starts media monitor when ice state is `connected`, and stops it when ice state is `closed` or `failed`. 

Checklist before merge
- [x] ~merge https://github.com/twilio/twilio-webrtc.js/pull/102, and update this change to use `timestamp`~ n/a
- [x] remove debug console.log calls, added in the change 
- [x] resolve and incorporate questions identified in https://paper.dropbox.com/doc/Ice-Media-Monitor-for-JS--AdD_RBac6xRW8YBvKcLizZ68AQ-z2sUflMxe4dXBrjfyWl42
- [x] add more unit tests to verify Monitor callbacks
- [ ] test for network change on chrome/firefox and safari

### CODE REVIEW

File: lib/signaling/v2/icemediamonitor.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  _getMediaStats() {
+    return this._pc.getMediaStats();
+  }
+  /**
+   * Start monitoring.
+   * @returns {void}
+   */
+  start(onInactiveCallback) {
```
Comment (makarandp0) [inline lib/signaling/v2/icemediamonitor.js]: I initially had  this class as EventEmitter (instead of taking callback param here), but decided against that, as callback works better in this case because 1) we have just one listener 2) the lifetime of callback is between start and stop, and it's easier for caller to call start/stop instead of an additional on/off for event listener.

File: test/unit/spec/signaling/v2/icemediamonitor.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      const monitor = new IceMediaMonitor(pc);
+      assert.equal(monitor._timer, null);
+      monitor.start(() => {});
+      assert.notEqual(monitor._timer, null);
+      monitor.stop();
+      assert.equal(monitor._timer, null);
+    });
+  });
+
+  describe('Callback', () => {
```
Comment (makarandp0) [inline test/unit/spec/signaling/v2/icemediamonitor.js]: WIP

File: lib/signaling/v2/icemediamonitor.js (hunk 1 above)
Comment (manjeshbhargav) [reply lib/signaling/v2/icemediamonitor.js]: @makarandp0 I think you can convert this to a Promise-based method and get rid of the `stop()` method 
 and the `_timer` property:
```js
class IceConnectionMonitor {
  ...
  /**
   * Check if the ICE connection is inactive.
   * @returns {Promise<void>}
   */
  checkIfInactive() {
    return wait(this._activityCheckPeriodMs).then(() => {
      return this._getIceConnectionStats();
    }).then(stats => {
      if (!stats) {
        return this.checkIfInactive();
      }
      if (!this._lastActiveStats || this._lastActiveStats.bytesReceived !== stats.bytesReceived) {
        this._lastActiveStats = stats;
      }
      if (stats.timestamp - this._lastActivity.timestamp < this._inactivityThresholdMs) {
        return this.checkIfInactive();
      }
    });
  }
  ...
}

// This can be moved to lib/utils/index.js

/**
 * Wait for a given period of time.
 * @param {number} delayMs
 * @returns {Promise<void>}
 */
function wait(delayMs) {
  return new Promise(resolve => setTimeout(resolve, delayMs));
}
```

File: lib/signaling/v2/icemediamonitor.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      _timer: {
+        value: null,
+        writable: true,
+      },
+      _activityCheckPeriodMS: {
+        value: options.activityCheckPeriodMS
+      },
+      _inactivityThresholdMS: {
+        value: options.inactivityThresholdMS
+      }
```
Comment (manjeshbhargav) [inline lib/signaling/v2/icemediamonitor.js]: @makarandp0 Can you define `_lastActiveStats` here?

File: lib/signaling/v2/peerconnection.js [hunk 4]
```diff
@@ -185,6 +187,13 @@ class PeerConnectionV2 extends StateMachine {
       _rtpSenders: {
         value: new Map()
       },
+      _mediaMonitorTimer: {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 Looks like this is not used anywhere.

File: lib/signaling/v2/peerconnection.js [hunk 5]
```diff
@@ -99,6 +100,7 @@ class PeerConnectionV2 extends StateMachine {
       setBitrateParameters,
       setCodecPreferences,
       setSimulcast,
+      IceMediaMonitor,
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 `IceConnectionMonitor`?

File: lib/signaling/v2/peerconnection.js [hunk 6]
```diff
@@ -7,6 +7,7 @@ const DefaultMediaStream = WebRTC.MediaStream;
 const DefaultRTCIceCandidate = WebRTC.RTCIceCandidate;
 const DefaultRTCPeerConnection = WebRTC.RTCPeerConnection;
 const DefaultRTCSessionDescription = WebRTC.RTCSessionDescription;
+const IceMediaMonitor = require('./icemediamonitor.js');
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 `iceconnectionmonitor.js`?

File: lib/signaling/v2/peerconnection.js [hunk 7]
```diff
@@ -185,6 +187,13 @@ class PeerConnectionV2 extends StateMachine {
       _rtpSenders: {
         value: new Map()
       },
+      _mediaMonitorTimer: {
+        writable: true,
+        value: null
+      },
+      _mediaMonitor: {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 Can you change this to:
```js
_iceConnectionMonitor: {
  value: new options.IceConnectionMonitor(peerConnection)
}
```
?

File: lib/signaling/v2/peerconnection.js [hunk 8]
```diff
@@ -474,10 +483,15 @@ class PeerConnectionV2 extends StateMachine {
   _handleIceConnectionStateChange() {
     const { iceConnectionState } = this._peerConnection;
 
-    this._log.debug(`ICE connection state is "${iceConnectionState}"`);
+    this._log.debug(`makarand: ICE connection state is "${this._lastIceConnectionState}" => "${iceConnectionState}"`);
+    // makarand: TODO: is it okay to short-circuit here? or anybody depends on
+    // iceConnectionStateChanged firing on every non-change
+    if (this._lastIceConnectionState === iceConnectionState) {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 Let's retain how it was before, unless you can demonstrate that "iceConnectionStateChanged" is emitted even when there is a non-change.

File: lib/signaling/v2/peerconnection.js [hunk 9]
```diff
@@ -488,10 +502,46 @@ class PeerConnectionV2 extends StateMachine {
       this._log.info('ICE reconnected');
     }
 
+    if (iceConnectionState === 'connected') {
+      this._mediaMonitor.start(() => {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 This can be:
```js
this._iceConnectionMonitor.checkIfInactive().then(() => {
  if (this._lastIceConnectionState !== 'disconnected') {
    return this._iceConnectionMonitor.checkIfInactive();
  }
  if (!this._shouldRestartIce && !this._isRestartingIce) {
    this._shouldRestartIce = true;
    this.offer();
  }
});
```

File: lib/signaling/v2/peerconnection.js [hunk 10]
```diff
@@ -488,10 +502,46 @@ class PeerConnectionV2 extends StateMachine {
...
+        this._log.debug(`makarand: Media Monitor detected inactivity while in "${this._lastIceConnectionState}" state`);
+        if (this._lastIceConnectionState === 'disconnected') {
+          this._mediaMonitor.stop();
+          if (!this._shouldRestartIce && !this._isRestartingIce) {
+            this._shouldRestartIce = true;
+            this.offer();
+          }
+        }
+      });
+    } else if (iceConnectionState === 'failed' || iceConnectionState === 'closed') {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 I don't think this will be necessary with the new way of checking inactivity, since we are using `setTimeout` instead of `setInterval`.

File: lib/signaling/v2/peerconnection.js [hunk 11]
```diff
@@ -488,10 +502,46 @@ class PeerConnectionV2 extends StateMachine {
...
+      this._mediaMonitor.stop();
+    }
+
     this._lastIceConnectionState = iceConnectionState;
     this.emit('iceConnectionStateChanged');
   }
 
+  // TODO: simplify and move to IceMediaMonitor once timestamp is added to the webrtc getStat api.
+  getMediaStats() {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 This method won't be necessary due to availability of `IceConnectionMonitor#_getIceConnectionStats()`.

File: lib/util/constants.js [hunk 12]
```diff
@@ -71,3 +71,6 @@ module.exports.typeErrors = {
 module.exports.DEFAULT_NQ_LEVEL_LOCAL = 1;
 module.exports.DEFAULT_NQ_LEVEL_REMOTE = 0;
 module.exports.MAX_NQ_LEVEL = 3;
+
+module.exports.ICE_MONITOR_ACTIVITY_CHECK_PERIOD_MS = 1000;
```
Comment (manjeshbhargav) [inline lib/util/constants.js]: @makarandp0 Nitpick: I think we can shorten this to `ICE_ACTIVITY_CHECK_PERIOD_MS`.

File: lib/util/constants.js [hunk 13]
```diff
@@ -71,3 +71,6 @@ module.exports.typeErrors = {
...
+module.exports.ICE_MONITOR_INACTIVITY_THRESHOLD_MS = 3000;
```
Comment (manjeshbhargav) [inline lib/util/constants.js]: @makarandp0 Nitpick: I think we can shorten this to `ICE_INACTIVITY_THRESHOLD_MS`.

File: test/unit/index.js [hunk 14]
```diff
@@ -49,6 +49,7 @@ require('./spec/signaling/v2/localparticipant');
 require('./spec/signaling/v2/networkqualitymonitor');
 require('./spec/signaling/v2/networkqualitysignaling');
 require('./spec/signaling/v2/recording');
+require('./spec/signaling/v2/icemediamonitor');
```
Comment (manjeshbhargav) [inline test/unit/index.js]: @makarandp0 `iceconnectionmonitor.js`.

File: lib/signaling/v2/icemediamonitor.js (hunk 1 above)
Comment (makarandp0) [reply lib/signaling/v2/icemediamonitor.js]: @manjeshbhargav, I think your suggestion would work if we decided to only start the monitor when Ice goes into disconnected state, and simply wait for 3 seconds of non-activity. But since we are monitoring at connected state, we need this callback to fire every time it detects inactivity, and peerConnection object acts on it only when it finds itself in disconnected state when callback fires. Maybe I can setup a quick meeting to syncup on this.

File: lib/signaling/v2/icemediamonitor.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
+'use strict';
+
+const { ICE_MONITOR_ACTIVITY_CHECK_PERIOD_MS, ICE_MONITOR_INACTIVITY_THRESHOLD_MS } = require('../../util/constants');
+
+/**
```
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/signaling/v2/icemediamonitor.js]: okay, will add line describing the class
 ```
  Monitors a {@link RTCPeerConnection}'s ICE media stats and notifies
 caller when inactivity is detected.
```
not mentioning the ICE connection state, as that logic is in PeerConnectionV2

File: lib/signaling/v2/icemediamonitor.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+ * @emits IceMediaMonitor#updated
+ */
+class IceMediaMonitor {
+  /**
+   * Construct a {@link IceMediaMonitor}.
+   */
+  constructor(pc, options) {
```
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/signaling/v2/icemediamonitor.js]: yes, that makes sense. I will update to use RTCPeerConnection, and also move the getMediaStats over to this file.

File: lib/signaling/v2/icemediamonitor.js [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  // TODO: remove this before merging.
+  _makarandlog() {
+    var newArgs = ['makarand: '];
+    var args = Array.prototype.slice.call(arguments);
+    console.log.apply(console.log, newArgs.concat(args));
+  }
+
+  _getMediaStats() {
```
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/signaling/v2/icemediamonitor.js]: Although it may not be necessary in this case - I thought using twilio-webrtc was preferred because it ensures standardized response across browsers. But I am open to using peerConnection directly.

File: lib/signaling/v2/peerconnection.js (hunk 8 above)
Comment (makarandp0) [reply lib/signaling/v2/peerconnection.js]: Looks like the code previously assumed that event could be emitted for non-changes. Notice that https://github.com/twilio/twilio-video.js/blob/master/lib/signaling/v2/peerconnection.js#L494  
explicitely ensures that old state is not "failed" 
```
    // Case 1: Transition to "failed".
    if (this._lastIceConnectionState !== 'failed' && iceConnectionState === 'failed' && !this._shouldRestartIce && !this._isRestartingIce) {
```

This change allows to get rid of that check.

File: lib/signaling/v2/peerconnection.js (hunk 9 above)
Comment (makarandp0) [reply lib/signaling/v2/peerconnection.js]: ah, Now I understand what you mean by using promise for this api -  Its using nested promises -  I think there are couple cons with this approach.

1. nested promises are making code bit complicated.
2. this usage makes it difficult to use "previous" state from the media monitor. For example handling of this case is not obvious.
a) connected: time 0 -  bytes = 0 - Api was called and it returned a promise..
b) connected: time 4 -  bytes = 0 - promise resolved. 
c) Api was was called again   because state is still connected.
d) disconnected:
e) disconnected: time 5 - bytes = 0.

At e) should the API resolve the promise? For callback it's obvious that it should fire again (as it was started, and callbacks on every inactivity). But since a new promise was started c) and it hasn't yet been 3 seconds of inactivity (although total inactivity is 5 seconds).


3. Also  with this promise API the promise never gets cleared (as there is no `stop` functionality) That means we will always have an outstanding promise (and associated timer) for peer connection - long after the peer connection is closed. This would lead to memory leaks, if there were lots of peerconnection on the app.

File: lib/signaling/v2/icemediamonitor.js (hunk 1 above)
Comment (makarandp0) [reply lib/signaling/v2/icemediamonitor.js]: never mind my comment above - I think I better understand your proposal now, and how it would work, but I do have few concerns with it, as I commented earlier.

File: lib/signaling/v2/peerconnection.js (hunk 10 above)
Comment (makarandp0) [reply lib/signaling/v2/peerconnection.js]: updated this now that we've decided to listen only after disconnect.

File: lib/signaling/v2/icemediamonitor.js (hunk 1 above)
Comment (makarandp0) [reply lib/signaling/v2/icemediamonitor.js]: updated this now that we've decided to listen only after disconnect.

File: lib/signaling/v2/peerconnection.js [hunk 18]
```diff
@@ -488,6 +493,19 @@ class PeerConnectionV2 extends StateMachine {
       this._log.info('ICE reconnected');
     }
 
+    if (iceConnectionState === 'disconnected') {
+      this._iceConnectionMonitor.start(() => {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 We need to make sure that when we access `PeerConnectionV2#iceConnectionState`, it is set to 'failed' here, even though the actual state is disconnected. We can do something like this:
```js
this._iceConnectionMonitor.start(() => {
  this._isIceConnectionInactive = true;
  ...
});
```
Then, we can modify the `iceConnectionState` getter for PeerConnectionV2 like this:
```js
get iceConnectionState() {
  return this._peerConnection.iceConnectionState === 'disconnected' && this._isIceConnectionInactive
    ? 'failed'
    : this._peerConnection.iceConnectionState;
}
```

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js [hunk 19]
```diff
@@ -0,0 +1,172 @@
+'use strict';
+
+const assert = require('assert');
+const sinon = require('sinon');
+
+const IceConnectionMonitor = require('../../../../../lib/signaling/v2/iceconnectionmonitor');
+
+describe('IceConnectionMonitor', () => {
+
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Nitpick: Remove extra line?

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+  describe('.constructor', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Nitpick: Remove `.`?

Comment (manjeshbhargav) [reply test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 I don't think we need to test for internal variables/methods unless absolutely required. Ideally we should just test public facing variables/methods like `start()` and `stop()`.

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+    it('defaults to 1sec of check period', () => {
+      assert.equal(new IceConnectionMonitor(pc)._activityCheckPeriodMs, 1000);
+    });
+
+    it('defaults to 3 sec of inactivity threshold', () => {
+      assert.equal(new IceConnectionMonitor(pc)._inactivityThresholdMs, 3000);
+    });
+  });
+
+  describe('.start', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Nitpick: `#start`?

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+    // eslint-disable-next-line no-undefined
+    ['foo', 45, { foo: 5 }, null].forEach((callback) => {
+      it(`throws if callback is ${typeof callback} `, () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Not needed.

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+    it('starts the timer', () => {
+      var pc = { foo: 1 };
+      const monitor = new IceConnectionMonitor(pc);
+      assert.equal(monitor._timer, null);
+      monitor.start(() => {});
+      assert.notEqual(monitor._timer, null);
+    });
+  });
+
+  describe('._getMediaStats', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Not necessary.

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+      it(`throws if callback is ${typeof callback} `, () => {
+        const pc = { foo: 1 };
+        const monitor = new IceConnectionMonitor(pc);
+        assert.throws(() => {
+          monitor.start(callback);
+        });
+      });
+    });
+
+    it('starts the timer', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 I think we can test the callback firing here, which is the only public facing behavior. We should not test internal implementation details like starting/stopping the timer.

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+      var pc = { foo: 1 };
+      const iceConnectionMonitor = new IceConnectionMonitor(pc);
+      sinon.stub(iceConnectionMonitor, '_getIceConnectionStats').returns(Promise.reject(new Error('boo')));
+      return iceConnectionMonitor._getMediaStats().then((mediaStats) => {
+        assert.strictEqual(mediaStats, null);
+      });
+    });
+  });
+
+  describe('.stop', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Nitpick: `#stop`?

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+    it('stops the timer, and resets the state', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Should not be testing implementation details, only expected behavior.

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+      assert.equal(monitor._timer, null);
+      monitor.start(() => {});
+      assert.notEqual(monitor._timer, null);
+      monitor.stop();
+      assert.equal(monitor._timer, null);
+      assert.equal(monitor._lastActivity, null);
+    });
+  });
+
+  describe('on inactivity', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 This can move to the `#start` test.

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js [hunk 28]
```diff
@@ -0,0 +1,172 @@
...
+
+      setTimeout(() => {
+        assert(monitor._getMediaStats.callCount < 12, 'timeout should reach before reasonable callCount');
+        done();
+      }, 10);
+    });
+  });
+});
+
+function mockMediaStats(iceConnectionMonitor, mediaStatResults) {
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/iceconnectionmonitor.js]: @makarandp0 `mockIceConnectionStats` and `statsResults`?

File: test/unit/spec/signaling/v2/peerconnection.js [hunk 29]
```diff
@@ -5,12 +5,34 @@ const EventEmitter = require('events');
...
+    });
+    sinon.stub(IceConnectionMonitor.prototype, 'stop').callsFake(() => {
+      didStopMonitor = true;
+    });
+  });
+  afterEach(() => {
+    IceConnectionMonitor.prototype.start.restore();
+    IceConnectionMonitor.prototype.stop.restore();
+  });
+
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/peerconnection.js]: @makarandp0 Can you also add or update the test for `PeerConnectionV2#iceConnectionState` to return "failed" if the actual state is "disconnected" and ICE is inactive, and should return the underlying RTCPeerConnection's `iceConnectionState` otherwise?

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+      return null;
+    });
+  }
+
+  /**
+   * Start monitoring.
+   * @returns {void}
+   */
+  start(onInactiveCallback) {
+    if (typeof onInactiveCallback !== 'function') throw new Error('invalid callback');
```
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/signaling/v2/iceconnectionmonitor.js]: @manjeshbhargav, I am big fan of asserts - but did not see usage in our codebase. this looks like something that we could assert for (and not have an elaborate throw like I have here). Do we use asserts in our code? basically some code that is mostly for for helping debug, catch bugs early ?

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 31]
```diff
@@ -0,0 +1,130 @@
...
+    return this._peerConnection.getStats().then(stats => Array.from(stats.values()).find(stat => {
+      return stat.type === 'candidate-pair' && stat.nominated;
+    }));
+  }
+
+  /**
+   * returns media stats or null.
+   * @returns Promise<?{timestamp, bytesReceived, bytesSent}>
+   */
+  _getMediaStats() {
```
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/signaling/v2/iceconnectionmonitor.js]: Re: introduction of _getMediaStats - I added it to  help isolate the peerConnection.getStats logic, and thus help write effective unit tests. like "getMediaStats works fine when _getIceConnectionStats throws or returns undefined"

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js (hunk 26 above)
Comment (makarandp0) [reply test/unit/spec/signaling/v2/iceconnectionmonitor.js]: it's a uni test, and I think it is useful to verify the internal state from unit test. But I agree overall, its better to test public interfaces. Here i did not want to add to the public interface to ensure that the timer is turned off.

File: lib/signaling/v2/iceconnectionmonitor.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/signaling/v2/iceconnectionmonitor.js]: I will remove this check, but wanted to discuss this with you more about.

File: test/unit/spec/signaling/v2/iceconnectionmonitor.js (hunk 25 above)
Comment (makarandp0) [reply test/unit/spec/signaling/v2/iceconnectionmonitor.js]: q: where do we use '.foo' v/s '#foo' ? didn't see any other usage of '#foo' in the code

File: test/unit/spec/signaling/v2/peerconnection.js (hunk 29 above)
Comment (makarandp0) [reply test/unit/spec/signaling/v2/peerconnection.js]: oh, I did not think about this - Do we need to keep track of the ice failure predicted by monitor and pretend that it was real failure to the callers of PeerConnectionV2 ? Let me sync up with you  understand this better.

**Comment by makarandp0** [comment]
Thank you for the feedback and catching unit test failure issue @manjeshbhargav - I think my tests were leaking a timer which was causing failures during other test runs ). I have fixed the issue, and added some instrumentation to see if it happens again. i will remove the instrumentation before merge along with other debug code.

File: lib/signaling/v2/peerconnection.js (hunk 18 above)
Comment (makarandp0) [reply lib/signaling/v2/peerconnection.js]: ah, I missed this comment. makes sense.

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 32]
```diff
@@ -0,0 +1,130 @@
+'use strict';
+
+const { ICE_ACTIVITY_CHECK_PERIOD_MS, ICE_INACTIVITY_THRESHOLD_MS } = require('../../util/constants');
+
+/**
+ * Monitors a {@link RTCPeerConnection}'s media stats and notifies
```
[the remark this answers is not in the document]
Comment (manjeshbhargav) [reply lib/signaling/v2/iceconnectionmonitor.js]: @makarandp0 "media" is still present.

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 33]
```diff
@@ -0,0 +1,128 @@
...
+    return this._peerConnection.getStats().then(stats => Array.from(stats.values()).find(stat => {
+      return stat.type === 'candidate-pair' && stat.nominated;
+    }));
+  }
+
+  /**
+   * returns ICE connection stats or null.
+   * @returns Promise<?{timestamp, bytesReceived, bytesSent}>
+   */
+  _getStats() {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/iceconnectionmonitor.js]: @makarandp0 I still think we don't need this method. Why would we want to copy the stats values into a JSON instead of just using the RTCIceCandidateStats directly? You can also attach the `catch()` handler to `_getIceConnectionStats()`.

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 34]
```diff
@@ -0,0 +1,128 @@
...
+        bytesReceived: activeCandidatePairStats.bytesReceived,
+        bytesSent: activeCandidatePairStats.bytesSent
+      };
+    }).catch(() => {
+      return null;
+    });
+  }
+
+  /**
+   * Start monitoring.
```
Comment (manjeshbhargav) [inline lib/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Can you add a more descriptive statement here along with specifying the parameter?
```js
/**
 * Start monitoring the ICE connection.
 * @param {function} onIceConnectionInactive
 * ...
 */
```

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 35]
```diff
@@ -0,0 +1,128 @@
...
+          onIceConnectionInactive();
+        }
+      }).catch((err) => {
+        this._makarandlog('error in _restartIceIfInactiveForLong', err);
+      });
+    }, this._activityCheckPeriodMs);
+  }
+
+  /**
+   * Stop monitoring.
```
Comment (manjeshbhargav) [inline lib/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Nitpick: `Stop monitoring the ICE connection state.`?

File: lib/signaling/v2/peerconnection.js [hunk 36]
```diff
@@ -290,7 +299,7 @@ class PeerConnectionV2 extends StateMachine {
    * @property {RTCIceConnectionState}
    */
   get iceConnectionState() {
-    return this._peerConnection.iceConnectionState;
+    return (this._isIceConnectionInactive && this._peerConnection.iceConnectionState === 'disconnected') ? 'failed' : this._peerConnection.iceConnectionState;
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 Nitpick: can you make this multiline so that its not too long?

File: lib/signaling/v2/peerconnection.js [hunk 37]
```diff
@@ -474,7 +483,7 @@ class PeerConnectionV2 extends StateMachine {
   _handleIceConnectionStateChange() {
     const { iceConnectionState } = this._peerConnection;
 
-    this._log.debug(`ICE connection state is "${iceConnectionState}"`);
+    this._log.debug(`ICE connection state is "${this._lastIceConnectionState}" => "${iceConnectionState}"`);
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 I think we should not change the log statement here since `_lastIceConnectionState` can be null.

File: lib/signaling/v2/iceconnectionmonitor.js (hunk 31 above)
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/signaling/v2/iceconnectionmonitor.js]: removed the method now.

File: lib/signaling/v2/iceconnectionmonitor.js (hunk 33 above)
Comment (makarandp0) [reply lib/signaling/v2/iceconnectionmonitor.js]: okay, I guess the method has served its purpose of being able to test. I will removed it and fold into `_getIceConnectionStats`

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 38]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  /**
+   * Get ICE connection stats, and extract received and send bytes.
+   * @returns Promise<?{timestamp, bytesReceived, bytesSent}>
+   */
+  _getIceConnectionStats() {
+    return this._peerConnection.getStats().then(stats => Array.from(stats.values()).find(stat => {
+      return stat.type === 'candidate-pair' && stat.nominated;
+    })).then((activeCandidatePairStats) => {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Nitpick: I think we can return the whole RTCIceCandidatePairStats here and get rid of the `then()` callback. Also, I think you can get rid of the braces around `activeIceCandidatePairStats`.

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 39]
```diff
@@ -0,0 +1,100 @@
...
+        // after inactivity threshold fire inactiveCallback.
+        if (iceStats.timestamp - this._lastActivity.timestamp >= this._inactivityThresholdMs) {
+          onIceConnectionInactive();
+        }
+      });
+    }, this._activityCheckPeriodMs);
+  }
+
+  /**
+   * Stop monitoring the ICE connection state
```
Comment (manjeshbhargav) [inline lib/signaling/v2/iceconnectionmonitor.js]: @makarandp0 Nitpick: Add a `.` at the end?

File: lib/signaling/v2/peerconnection.js [hunk 40]
```diff
@@ -290,7 +299,8 @@ class PeerConnectionV2 extends StateMachine {
    * @property {RTCIceConnectionState}
    */
   get iceConnectionState() {
-    return this._peerConnection.iceConnectionState;
+    return (this._isIceConnectionInactive && this._peerConnection.iceConnectionState === 'disconnected')
```
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js]: @makarandp0 Nitpick: I don't think we need the `()` around the condition. Also, can you indent the `?` with just two spaces after `return` like this:
```js
return this._isIceConnectionInactive && this._peerConnection.iceConnectionState === 'disconnected'
  ? 'failed' : this._peerConnection.iceConnectionState;
```
?

File: lib/signaling/v2/iceconnectionmonitor.js (hunk 38 above)
Comment (makarandp0) [reply lib/signaling/v2/iceconnectionmonitor.js]: imo having a limited fields returned from _getIceConnectionStats, helps with unit test/mocking of this function. I will  update  to get rid of braces.

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 41]
```diff
@@ -46,13 +46,7 @@ class IceConnectionMonitor {
...
     return this._peerConnection.getStats().then(stats => Array.from(stats.values()).find(stat => {
       return stat.type === 'candidate-pair' && stat.nominated;
-    })).then((activeCandidatePairStats) => {
-      return {
-        timestamp: activeCandidatePairStats.timestamp,
-        bytesReceived: activeCandidatePairStats.bytesReceived,
-        bytesSent: activeCandidatePairStats.bytesSent
-      };
-    }).catch(() => {
+    })).catch(() => {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/iceconnectionmonitor.js]: @makarandp0 I'll leave it up to you, but the catch callback can also be the second argument for `then()`:
```js
.then(onResolved, onRejected);
```

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 42]
```diff
@@ -0,0 +1,94 @@
...
+      _timer: {
+        value: null,
+        writable: true,
+      }
+    });
+  }
+
+  /**
+   * Get ICE connection stats, and extract received and send bytes.
+   * @returns Promise<?{timestamp, bytesReceived, bytesSent}>
```
Comment (manjeshbhargav) [inline lib/signaling/v2/iceconnectionmonitor.js]: @makarandp0 `@returns Promise<?RTCIceCandidatePairStats>`?

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 43]
```diff
@@ -0,0 +1,94 @@
...
+   */
+  start(onIceConnectionInactive) {
+    this.stop();
+    this._timer = setInterval(() => {
+      this._getIceConnectionStats().then(iceStats => {
+        if (!iceStats) {
+          return;
+        }
+
+        // if change detected in bytesReceived mark it as lastActiveiy.
```
Comment (manjeshbhargav) [inline lib/signaling/v2/iceconnectionmonitor.js]: @makarandp0 I think the code is self-explanatory, so we don't need the comment here.

File: lib/signaling/v2/iceconnectionmonitor.js [hunk 44]
```diff
@@ -0,0 +1,94 @@
...
+        if (!this._lastActivity || this._lastActivity.bytesReceived !== iceStats.bytesReceived) {
+          this._lastActivity = iceStats;
+        }
+
+        // after inactivity threshold fire inactiveCallback.
```
Comment (manjeshbhargav) [inline lib/signaling/v2/iceconnectionmonitor.js]: @makarandp0 I think the code is self-explanatory, so we don't need the comment here.

File: lib/signaling/v2/iceconnectionmonitor.js (hunk 41 above)
Comment (makarandp0) [reply lib/signaling/v2/iceconnectionmonitor.js]: @manjeshbhargav, I do see usage of 2nd param of `.then` in our codebase, and thinking we should move to using .catch instead - 1st type of usage could be problematic. More about this antipattern at https://www.datchley.name/promise-patterns-anti-patterns/
